[Guide] add guide pages for views/modals (#7507)
Rapptz/discord.py: docs/guide/interactions/modals.rst

.. currentmodule:: discord

.. _guide_interactions_modals:

Modals
========

Modals are a form of interaction which prompt the user for additional information, they appear as a pop-up windows and accept input using
components such as :class:`~discord.ui.TextInput`.

This section will detail how to create and use these modals in your code.

Defining a Modal
------------------

> Modals share structural similarities with :ref:`Views <guide_interactions_views>`, both are used to design component based UIs.

Reviewer: ```suggestion
Modals share structural similarities with :ref:`Views <guide_interactions_views>`, both are used to design component-based UIs.
```

> and share 

Reviewer 2: ```suggestion
and share 
```

and share what

Modals are created by subclassing the :class:`~discord.ui.Modal` class.

> For example, the following class definition:

Reviewer 2: ```suggestion
For example:
```

continued below

> .. code-block:: python3

Reviewer 2: ```suggestion
.. code-block:: python
```

deprecated i think

    class FeedbackForm(discord.ui.Modal, title="Feedback"):
        name = discord.ui.TextInput(label="Name")
        feedback = discord.ui.TextInput(label="Feedback", style=discord.TextStyle.long)
        additional_information = discord.ui.TextInput(
            label="Additional Information", style=discord.TextStyle.long, required=False
        )

        async def on_submit(self, interaction: discord.Interaction) -> None:
            ... # do something with the data

            await interaction.response.send_message(f'Thank you {self.name.value}, your submission was recorded.')

> produces a modal which appears on Discord as:

Reviewer 2: ```suggestion
This produces a modal which appears on Discord as:
```

probably don't continue sentences around blocks of code

.. image:: /images/guide/interactions/modal_example.png

Let's break down the code:

> Title

Reviewer 3: No brief example?

Reviewer 3: I might have misread, and was referring to what a modal does. You can mark this resolved

Title
~~~~~~

When defining the class a ``title`` keyword argument can be passed which sets the default title of the modal.
Instances of the modal class can override this title by setting the ``title`` attribute or passing a ``title`` keyword argument to

> the constructor. Titles are required and are displayed at the top of the modal.

Reviewer 2: ```suggestion
the constructor. The title is required and is displayed at the top of the modal.
```

one title

Fields
~~~~~~~

In our example above we have three class-attributes: ``name``, ``feedback``, and ``additional_information``, all of which are instances of the :class:`~discord.ui.TextInput` class.
In modal classes, class-attributes which are instances of components are used to represent the fields of the modal.

Discord requires a modal contain between 1 and 5 fields. 

.. note:: 

>     Each modal fields is displayed in definition order. Our example from above defines ``name`` before it defines ``feedback``, 

Reviewer 2: ```suggestion
    Each modal field is displayed in definition order. Our example from above defines ``name`` before it defines ``feedback``, 
```

one fields

    that order is preserved when the FeedbackForm modal is displayed on Discord.

> Fields can also be added or removed from a modal instance using the :meth:`~discord.ui.View.add_item` and :meth:`~discord.ui.View.remove_item` methods.

Reviewer 2: ```suggestion
Fields can also be added to or removed from a modal instance using the :meth:`~discord.ui.View.add_item` and :meth:`~discord.ui.View.remove_item` methods.
```

reads nicer

.. warning:: 

>     Ensure that class-attribute defined fields do not conflict with attributes or methods defined in the class, as this

Reviewer 2: probably tell me what attributes these are and where

    will cause unexpected behavior.

> You can customize individual modal instances with normal attribute access and assignment, For example, to change the label of the ``name`` field to ``"Your Name"``:

Reviewer 2: ```suggestion
You can customize individual modal instances with normal attribute access and assignment, For example, to change the label of the ``name`` field to ``"Your Name"`` you can do the following:
```

i don't love "you can do the following" but it needs something

.. code-block:: python3

    form = FeedbackForm()
    form.name.label = "Your Name"

Handling User Responses
~~~~~~~~~~~~~~~~~~~~~~~~~

Once the user has clicked `Submit` the :meth:`Modal.on_submit <discord.ui.Modal.on_submit>` callback is called. 

> The callback is passed a new :class:`~discord.Interaction` instance, which requires a response.

Reviewer 2: what if i don't respond? what is the timeout?

In the example above we respond by sending a message to the user confirming their submission was recorded.

Handling an error
~~~~~~~~~~~~~~~~~~

The :meth:`Modal.on_error <discord.ui.Modal.on_error>` method is called when an exception is raised
within :meth:`Modal.on_submit <discord.ui.Modal.on_submit>`, and can be used to handle an error or respond to the user.

This method is passed the exception raised, and the :class:`~discord.Interaction` instance, allowing you to handle the error and, 
if necessary, send a response to the user.

.. code-block:: python
    :emphasize-lines: 13-15

    class FeedbackForm(discord.ui.Modal, title='Feedback'):
        name = discord.ui.TextInput(label='Name')
        feedback = discord.ui.TextInput(label='Feedback', style=discord.TextStyle.long)
        additional_information = discord.ui.TextInput(
            label='Additional Information', style=discord.TextStyle.long, required=False
        )

        async def on_submit(self, interaction: discord.Interaction) -> None:
            ... # do something with the data

            await interaction.response.send_message(f'Thank you {self.name.value}, your submission was recorded.')

>         async def on_error(self, error: Exception, interaction: discord.Interaction) -> None:

Reviewer 3: The arguments for `on_error` are incorrect. The proper ones should be noted below:

`(self, interaction: discord.Interaction, error: Exception)`

>             if not interaction.response.is_done():

Reviewer 3: As the interaction will be sent once, there isn't a need to set a guard for it

Reviewer 3: The guard will **always** be ran, so it's unnecessary

Reviewer 4: Agreed on it being best practice, I would disregard this comment.

                await interaction.response.send_message('An error occurred, please try again.')

Sending a Modal
-----------------

A modal can be sent by calling the :meth:`InteractionResponse.send_modal` method when handling an :class:`Interaction`.

For example, you can respond with a modal when somebody uses a slash command:

.. code-block:: python

    @discord.app_commands.command()
    async def feedback(interaction: discord.Interaction) -> None:
        """Send important feedback."""
        await interaction.response.send_modal(FeedbackForm())

Once a modal is sent, a user has the option to submit values. These should be handled either in an implementation of the

> :meth:`~discord.ui.Modal.on_submit` method or via :meth:`Modal.wait() <discord.ui.Modal.wait>`,

Reviewer 3: Kinda on the fence about including `Modal.wait` usage as I've never seen code that uses the coro

Reviewer 3: Well if that works, then it's fine to leave it there

in a similar fashion to the :ref:`confirmation prompt View example <guide_interactions_views>`.

.. note::

>     It is currently not possible to tell if a user has closed the modal prompt without submitting.

Reviewer 3: It has never been to my knowledge
